rfcs/0000-simple_unwind_annotation.md

```diff
@@ -1,179 +1,311 @@
+(recall, the `p` in `pthreads` stands for POSIX).
+
+[x86-64 ps-ABI]: https://github.com/hjl-tools/x86-psABI/wiki/X86-psABI
+[itanium]: https://itanium-cxx-abi.github.io/cxx-abi/abi.html
+[pthreads]: http://man7.org/linux/man-pages/man7/pthreads.7.html
+
+## Unwinding from a Rust callback in C: "C unwind" calls Rust
+
+Some C libraries like `mozjpeg` take callbacks using the platform native ABI and
+expect these callbacks to unwind through C back into the user application to
+report errors.
+
+JIT compilers like [Lucet][lucet] and [Weld][weld]: TODO (I'm not sure I understood what these programs do; need some help here)
+  * they generate code that calls other code using the C ABI
+  * all the code except the ABI is implemented in Rust, so they want to be able to unwind through the ABI
+  * the `"Rust"` call ABI is too "unstable" for their uses, and they prefer to use the more stable platform ABI
+
+[lucet]: https://github.com/fastly/lucet	
 [weld]: https://www.weld.rs/
 
-## Less dependence on C/C++
-
-Some of the [proposed workarounds][cffi-panic] to the lack of cross-FFI unwinding require the use of
-wrappers written in C (for `setjmp` and `longjmp`) or C++ (for exceptions). This solution reduces
-the amount of non-Rust code that must be generated and maintained when the foreign code is
-compatible with the unspecified Rust unwinding mechanism.
-
-[cffi-panic]: https://github.com/gnzlbg/cffi-panic
-
 # Guide-level explanation
 [guide-level-explanation]: #guide-level-explanation
 
-Rust function definitions with the `"C panic"` ABI string (e.g., `extern "C panic" fn`) are
-permitted to unwind with a panic, as opposed to `extern "C" fn` functions which will abort the
-process if a panic reaches the function boundary.
-
-When used on declarations of imported functions (e.g., `extern "C panic" { fn ... }`) or function
-pointers (e.g., `extern "C panic" fn()`), the `"C panic"` ABI string means that if the function
-unwinds, the unwind will be propagated though any calling code. If an `extern "C"` imported function
-or function pointer unwinds, the behavior is undefined.
+This RFC introduces a new ABI string called `"C unwind"` on some specific
+targets. If this ABI is not available for a target, Rust will error at
+compile-time saying so. This RFC adds support of this ABI for most targets with
+unwinding support, but it is expected that support for this ABI will be added to
+more targets in the future, either in other RFCs that build on this one, or in
+FCPs.
+
+Like for all other ABI strings, one can use this ABI string:
+
+  * to define a function in Rust: `extern "C unwind" fn foo() { }`
+  * to declare a function in Rust: `extern "C unwind" { fn foo(); }`
+  * in function pointer types: `let _: extern "C unwind" fn = foo;`
+    * Since `"C unwind"` is a super-set of `"C"`, there are implicit coercions
+      from `extern "C"` function types to `extern "C unwind"` function types.
```

Comment:
I would prefer not to have such coercions, in part just to keep things simple. I think they are a fine bit of potential future work, though. That said, I guess it wouldn't be too hard to implement -- I wasn't sure if we would want to guarantee that "C" and "C unwind" are interoperable. Probably we simply do, in which case such coercions seem harmless enough.

Reply:
I agree. I don't think the coercions are required for an MVP, I just had the rules already written somewhere and c&p them here. I think I'm going to move this to the future section. If someone were to implement this before the PR for stabilization, one can re-consider at that point whether we want to stabilize them as well or leave them as unstable, but they do not need to block the stabilization of this feature.